File: TCR24CS020 - Catherine Maria Benny/tests/test_cases.cpp
// Unit tests for the Safe Semantic Planner (assignment section 31).
// Minimal, dependency-free assert-based test runner -- no external test
// framework required, matching the "do not overengineer" guidance (section 41).
//
// Covers: basic reachability, unreachable goal, bad initial state, bad goal,
// bad-state avoidance, unavailable transition, transition addition,
// transition removal, goal update, safety-distance calculation, path
// validation, and dynamic replanning.
#include <cmath>
#include <functional>
#include <iostream>
#include <string>
#include <vector>

#include "DStarLite.h"
#include "Metrics.h"
#include "TestScenarios.h"

namespace {
int g_passed = 0;
int g_failed = 0;

void check(bool cond, const std::string& testName, const std::string& detail = "") {
    if (cond) {
        ++g_passed;
        std::cout << "[PASS] " << testName << "\n";
    } else {
        ++g_failed;
        std::cout << "[FAIL] " << testName;
        if (!detail.empty()) std::cout << " -- " << detail;
        std::cout << "\n";
    }
}
} // namespace

static void test_BasicReachability() {
    DStarLite planner;
    auto r = planner.plan(scenarios::testCase1_BasicReachability());
    check(r.success, "BasicReachability: success");
    check(r.statePath == std::vector<uint64_t>({0, 1, 2, 3}), "BasicReachability: correct path");
    check(r.totalCost == 3.0, "BasicReachability: correct cost");
}

static void test_UnreachableGoal() {
    PlanningProblem p;
    p.states = {State(0, {0, 0}), State(1, {5, 5})};
    // No transitions at all -> goal unreachable.
    p.initialState = 0;
    p.goalState = 1;
    DStarLite planner;
    auto r = planner.plan(p);
    check(!r.success, "UnreachableGoal: fails cleanly", r.success ? "returned success" : "");
    check(!r.errorMessage.empty(), "UnreachableGoal: has an error message");
}

static void test_BadInitialState() {
    PlanningProblem p;
    p.states = {State(0, {0, 0}), State(1, {1, 0})};
    p.transitions = {Transition(0, 0, 1, 1.0, 0.9, 0.9, true)};
    p.initialState = 0;
    p.goalState = 1;
    p.badStates = {0}; // initial state itself is bad
    DStarLite planner;
    auto r = planner.plan(p);
    check(!r.success, "BadInitialState: planning fails");
    check(r.errorMessage.find("initial") != std::string::npos,
          "BadInitialState: error mentions initial state", r.errorMessage);
}

static void test_BadGoalState() {
    PlanningProblem p;
    p.states = {State(0, {0, 0}), State(1, {1, 0})};
    p.transitions = {Transition(0, 0, 1, 1.0, 0.9, 0.9, true)};
    p.initialState = 0;
    p.goalState = 1;
    p.badStates = {1}; // goal itself is bad
    DStarLite planner;
    auto r = planner.plan(p);
    check(!r.success, "BadGoalState: planning fails");
    check(r.errorMessage.find("goal") != std::string::npos,
          "BadGoalState: error mentions goal state", r.errorMessage);
}

static void test_BadStateAvoidance() {
    DStarLite planner;
    auto r = planner.plan(scenarios::testCase2_BadStateAvoidance());
    check(r.success, "BadStateAvoidance: success");
    check(r.badStatesVisited == 0, "BadStateAvoidance: zero bad states visited");
    bool visitsX = false;
    for (auto s : r.statePath) if (s == 2) visitsX = true;
    check(!visitsX, "BadStateAvoidance: bad state X never appears in path");
}

static void test_UnavailableTransition() {
    auto problem = scenarios::testCase4_DynamicTransition();
    DStarLite planner;
    auto initial = planner.plan(problem);
    check(initial.success, "UnavailableTransition: initial plan succeeds");
    auto replanned = planner.setTransitionAvailability(1, false);
    check(replanned.success, "UnavailableTransition: replan after disabling edge succeeds");
    bool usesDisabledEdge = false;
    for (auto tid : replanned.transitionPath) if (tid == 1) usesDisabledEdge = true;
    check(!usesDisabledEdge, "UnavailableTransition: new path avoids the disabled transition");
}

static void test_TransitionAddition() {
    auto problem = scenarios::testCase6_TransitionAddition();
    DStarLite planner;
    auto initial = planner.plan(problem);
    check(initial.success && initial.totalCost == 4.0,
          "TransitionAddition: initial detour cost is 4.0");
    Transition shortcut(10, 0, 3, 3.0, 0.9, 0.9, true);
    auto replanned = planner.addTransition(shortcut);
    check(replanned.success, "TransitionAddition: replan succeeds");
    check(replanned.totalCost == 3.0, "TransitionAddition: shortcut is adopted (cost 3.0)");
}

static void test_TransitionRemoval() {
    auto problem = scenarios::testCase4_DynamicTransition();
    DStarLite planner;
    planner.plan(problem);
    auto afterRemoval = planner.removeTransition(1); // A->G removed entirely
    check(afterRemoval.success, "TransitionRemoval: alternate route still found");
    check(afterRemoval.totalCost > 4.0, "TransitionRemoval: alternate route costs more than the original");
}

static void test_GoalUpdate() {
    auto problem = scenarios::testCase5_GoalUpdate();
    DStarLite planner;
    auto initial = planner.plan(problem);
    check(initial.success && initial.statePath.back() == 2, "GoalUpdate: initial plan reaches G1");
    auto replanned = planner.updateGoal(5);
    check(replanned.success && replanned.statePath.back() == 5, "GoalUpdate: replanned path reaches G2");
}

static void test_SafetyDistanceCalculation() {
    std::unordered_map<uint64_t, State> states;
    states[0] = State(0, {0.0, 0.0});
    states[1] = State(1, {3.0, 4.0}); // distance 5 from origin
    states[2] = State(2, {10.0, 0.0});
    std::vector<uint64_t> path = {0, 1, 2};
    std::vector<uint64_t> bad = {2};
    double d = metrics::minimumSafetyDistance(path, bad, states);
    // min( dist(0,2)=10, dist(1,2)=sqrt(49+16)=8.062, dist(2,2)=0 ) = 0
    check(std::abs(d - 0.0) < 1e-9, "SafetyDistanceCalculation: min distance is 0 when path visits the bad state",
          std::to_string(d));

    std::vector<uint64_t> path2 = {0, 1};
    double d2 = metrics::minimumSafetyDistance(path2, bad, states);
    double expected = std::min(metrics::euclideanDistance(states[0].embedding, states[2].embedding),
                                metrics::euclideanDistance(states[1].embedding, states[2].embedding));
    check(std::abs(d2 - expected) < 1e-9, "SafetyDistanceCalculation: correct min distance over a path",
          std::to_string(d2) + " vs " + std::to_string(expected));

    double dNoBad = metrics::minimumSafetyDistance(path2, {}, states);
    check(std::isinf(dNoBad), "SafetyDistanceCalculation: no bad states => +infinity (documented convention)");
}

static void test_PathValidation() {
    auto problem = scenarios::testCase1_BasicReachability();
    std::string err = metrics::validatePath(problem, {0, 1, 2, 3}, {0, 1, 2});
    check(err.empty(), "PathValidation: a correct path validates cleanly", err);

    std::string err2 = metrics::validatePath(problem, {1, 2, 3}, {1, 2});
    check(!err2.empty(), "PathValidation: rejects a path not starting at the initial state");

    std::string err3 = metrics::validatePath(problem, {0, 1, 2}, {0, 1});
    check(!err3.empty(), "PathValidation: rejects a path not ending at the goal");

    auto problem2 = scenarios::testCase2_BadStateAvoidance();
    std::string err4 = metrics::validatePath(problem2, {0, 1, 2, 4}, {0, 1, 2});
    check(!err4.empty(), "PathValidation: rejects a path that visits a bad state", err4);
}

static void test_DynamicReplanningReusesSearch() {
    auto problem = scenarios::testCase4_DynamicTransition();
    DStarLite planner;
    auto initial = planner.plan(problem);
    auto replanned = planner.setTransitionAvailability(1, false);
    // A correct incremental implementation only needs to re-expand the
    // handful of vertices in the affected region, not the whole graph again.
    check(replanned.exploredStates <= static_cast<int>(problem.states.size()),
          "DynamicReplanning: replanning explores a bounded number of vertices",
          std::to_string(replanned.exploredStates));
    check(replanned.success, "DynamicReplanning: still finds a valid alternate path");
}

int main() {
    std::cout << "Running Safe Semantic Planner unit tests...\n\n";
    test_BasicReachability();
    test_UnreachableGoal();
    test_BadInitialState();
    test_BadGoalState();
    test_BadStateAvoidance();
    test_UnavailableTransition();
    test_TransitionAddition();
    test_TransitionRemoval();
    test_GoalUpdate();
    test_SafetyDistanceCalculation();
    test_PathValidation();
    test_DynamicReplanningReusesSearch();

    std::cout << "\n" << g_passed << " passed, " << g_failed << " failed.\n";
    return g_failed == 0 ? 0 : 1;
}

File: TCR24CS065 - Sooraj K R/test_cases.py
# test_cases.py
# All 6 test cases from the assignment PDF.
# Each function creates and returns a PlanningProblem.

from structures import State, Transition, PlanningProblem


# ============================================================
# Test Case 1: Basic Reachability
# Simple linear path: S -> A -> B -> G
# No bad states. Should find the only valid path.
# ============================================================
def create_test_case_1():
    states = [
        State(id=0, embedding=[0.0, 0.0]),   # S (start)
        State(id=1, embedding=[1.0, 0.0]),   # A
        State(id=2, embedding=[2.0, 0.0]),   # B
        State(id=3, embedding=[3.0, 0.0]),   # G (goal)
    ]

    transitions = [
        Transition(id=10, from_state=0, to_state=1, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=11, from_state=1, to_state=2, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=12, from_state=2, to_state=3, cost=1.0, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=3,
        badStates=[],
        states=states,
        transitions=transitions
    )


# ============================================================
# Test Case 2: Bad State Avoidance
# Path 1: S -> A -> X -> G  (X is a bad state!)
# Path 2: S -> C -> D -> G  (safe path)
# Planner should pick Path 2.
# ============================================================
def create_test_case_2():
    states = [
        State(id=0, embedding=[0.0, 0.0]),    # S (start)
        State(id=1, embedding=[1.0, 1.0]),    # A
        State(id=2, embedding=[2.0, 1.0]),    # X (BAD STATE)
        State(id=3, embedding=[1.0, -1.0]),   # C
        State(id=4, embedding=[3.0, 0.0]),    # G (goal)
        State(id=5, embedding=[2.0, -1.0]),   # D
    ]

    transitions = [
        # Path 1 (through bad state X)
        Transition(id=20, from_state=0, to_state=1, cost=1.0, safety=0.5, reliability=1.0, available=True),
        Transition(id=21, from_state=1, to_state=2, cost=1.0, safety=0.0, reliability=1.0, available=True),
        Transition(id=22, from_state=2, to_state=4, cost=1.0, safety=0.0, reliability=1.0, available=True),
        # Path 2 (safe detour)
        Transition(id=23, from_state=0, to_state=3, cost=1.5, safety=1.0, reliability=1.0, available=True),
        Transition(id=24, from_state=3, to_state=5, cost=1.5, safety=1.0, reliability=1.0, available=True),
        Transition(id=25, from_state=5, to_state=4, cost=1.5, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=4,
        badStates=[2],       # X is bad
        states=states,
        transitions=transitions
    )


# ============================================================
# Test Case 3: Safety Margin
# Two valid paths, neither goes through a bad state, but:
# - Path 1: cheaper cost, but passes CLOSE to bad states
# - Path 2: higher cost, but stays FAR from bad states
# This tests how the planner balances cost vs safety distance.
# ============================================================
def create_test_case_3():
    states = [
        State(id=0, embedding=[0.0, 0.0]),    # S (start)
        State(id=1, embedding=[1.0, 0.5]),    # A (close to bad state B1)
        State(id=2, embedding=[2.0, 0.5]),    # B (close to bad state B1)
        State(id=3, embedding=[1.0, -3.0]),   # C (far from bad states)
        State(id=4, embedding=[2.0, -3.0]),   # D (far from bad states)
        State(id=5, embedding=[3.0, 0.0]),    # G (goal)
        State(id=6, embedding=[1.0, 1.0]),    # B1 (bad state, near path 1)
    ]

    transitions = [
        # Path 1: S -> A -> B -> G  (cheap but close to bad state)
        Transition(id=30, from_state=0, to_state=1, cost=1.0, safety=0.5, reliability=0.9, available=True),
        Transition(id=31, from_state=1, to_state=2, cost=1.0, safety=0.5, reliability=0.9, available=True),
        Transition(id=32, from_state=2, to_state=5, cost=1.0, safety=0.5, reliability=0.9, available=True),
        # Path 2: S -> C -> D -> G  (expensive but far from bad states)
        Transition(id=33, from_state=0, to_state=3, cost=2.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=34, from_state=3, to_state=4, cost=2.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=35, from_state=4, to_state=5, cost=2.0, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=5,
        badStates=[6],       # B1 is the bad state
        states=states,
        transitions=transitions
    )


# ============================================================
# Test Case 4: Dynamic Transition
# Initially: S -> A -> G (direct path)
# Also:      S -> B -> C -> G (backup path)
# After planning, transition A -> G becomes unavailable.
# Planner should find the backup path.
# ============================================================
def create_test_case_4():
    states = [
        State(id=0, embedding=[0.0, 0.0]),    # S (start)
        State(id=1, embedding=[1.0, 0.0]),    # A
        State(id=2, embedding=[1.0, -1.0]),   # B
        State(id=3, embedding=[2.0, -1.0]),   # C
        State(id=4, embedding=[3.0, 0.0]),    # G (goal)
    ]

    transitions = [
        # Direct path: S -> A -> G
        Transition(id=40, from_state=0, to_state=1, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=41, from_state=1, to_state=4, cost=1.0, safety=1.0, reliability=1.0, available=True),
        # Backup path: S -> B -> C -> G
        Transition(id=42, from_state=0, to_state=2, cost=1.5, safety=1.0, reliability=1.0, available=True),
        Transition(id=43, from_state=2, to_state=3, cost=1.5, safety=1.0, reliability=1.0, available=True),
        Transition(id=44, from_state=3, to_state=4, cost=1.5, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=4,
        badStates=[],
        states=states,
        transitions=transitions
    )


# ============================================================
# Test Case 5: Goal Update
# S -> A -> G1 (original goal)
# S -> A -> B -> G2 (new goal after update)
# The goal changes from G1 to G2, planner should find new path.
# ============================================================
def create_test_case_5():
    states = [
        State(id=0, embedding=[0.0, 0.0]),    # S (start)
        State(id=1, embedding=[1.0, 0.0]),    # A
        State(id=2, embedding=[2.0, 0.0]),    # G1 (original goal)
        State(id=3, embedding=[2.0, 1.0]),    # B
        State(id=4, embedding=[3.0, 1.0]),    # G2 (new goal)
    ]

    transitions = [
        Transition(id=50, from_state=0, to_state=1, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=51, from_state=1, to_state=2, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=52, from_state=1, to_state=3, cost=1.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=53, from_state=3, to_state=4, cost=1.0, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=2,         # original goal is G1
        badStates=[],
        states=states,
        transitions=transitions
    )


# ============================================================
# Test Case 6: Transition Addition
# Initially: S -> A -> B -> G (long path)
# Later: a new shortcut S -> G is added
# Planner should discover the improved solution.
# ============================================================
def create_test_case_6():
    states = [
        State(id=0, embedding=[0.0, 0.0]),    # S (start)
        State(id=1, embedding=[1.0, 0.0]),    # A
        State(id=2, embedding=[2.0, 0.0]),    # B
        State(id=3, embedding=[3.0, 0.0]),    # G (goal)
    ]

    transitions = [
        # Original long path: S -> A -> B -> G
        Transition(id=60, from_state=0, to_state=1, cost=2.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=61, from_state=1, to_state=2, cost=2.0, safety=1.0, reliability=1.0, available=True),
        Transition(id=62, from_state=2, to_state=3, cost=2.0, safety=1.0, reliability=1.0, available=True),
    ]

    return PlanningProblem(
        initialState=0,
        goalState=3,
        badStates=[],
        states=states,
        transitions=transitions
    )
